cha/implementaion/multidevice/our_communication.tex

```diff
@@ -0,0 +1,197 @@
+\cref{fig:connecting} we show the two new UIs made for connecting devices.
+\cref{fig:mode_selection} is the UI the app starts with.
+When in this state the user chooses to be either a server or a client.
+\cref{fig:group_selection} presents the UI a user is shown when pressing the ``Client'' option.
+It contains a list of all nearby devices which are acting as a ``Server'', how this is achieved is explained next.
+
+\begin{figure}[ht]
+  \begin{subfigure}[b]{0.33\linewidth}
+    \centering
+    %\includegraphics[width=0.75\linewidth]{example-image-a}
+    \includegraphics[trim={0cm 0cm 0cm 0cm}, clip, height=7cm]{img/ui/mode_selection.png}
+    \caption{Mode Selection}
+    \label{fig:mode_selection}
+    \vspace{4ex}
+  \end{subfigure}%%
+  \begin{subfigure}[b]{0.33\linewidth}
+    \centering
+    \includegraphics[trim={0cm 0cm 0cm 0cm}, clip, height=7cm]{img/ui/group_selection.png}
+    \caption{Group Selection}
+    \label{fig:group_selection}
+    \vspace{4ex}
+  \end{subfigure}%%
+  \begin{subfigure}[b]{0.33\linewidth}
+    \centering
+    \includegraphics[trim={0cm 0cm 0cm 0cm}, clip, height=7cm]{img/ui/wifi_direct_invitation.png}
+    \caption{WiFi Direct invitation}
+    \label{fig:wifidirectinv}
+    \vspace{4ex}
+  \end{subfigure}
+  \caption{The user interface relevant for connecting devices.}
+  \label{fig:connecting}
+\end{figure}
+
+\subsection{Service Discovery}
+
+The first problem which must be addressed in order to connect devices is peer discovery.
+Fortunately a part of the WiFi Direct standard is the so called services.
+Fortunately WiFi Direct has a number of helpful services.
+One of these services is ``Service Discovery'', which is made for the exact purpose of discovering peers.
+In Android WiFi Direct is called: ``Wi-Fi peer-to-peer'' or ``Wi-Fi P2P'' for short.\footnote{\url{https://developer.android.com/reference/android/net/wifi/p2p/WifiP2pManager.html}}.
+
+We experienced several issues implementing Service Discovery in our app, the most notable being group negotiation.
+When devices connect using WiFi Direct, they negotiate about having the role of group owner.
+The group owner is the master of the network, and is responsible for granting new devices access etc.
+For our app we want the device which is the server, to also have the role of being the group owner of the WiFi Direct network.
+The root of the problem was caused by the devices keeping a record of the last 32 WiFi Direct groups which they had been a part of and who were the group owner.
+This becomes a problem, when a new session using the same devices as a previous one, but with a different master is started.
+
+In order to prevent this issue, we decided to erase the records of previous WiFi Direct networks, when we start discovering services.
+In addition to the server creating a new group, once it starts broadcasting its service.
```

Comment:
This sentence makes no sense, in addition to the server creating a new group it does what when it starts broadcasting?